### ISSUE TITLE
caddy: setting configFile doesn't work

### ISSUE THREAD

**Issue by flokli** [NixOS/nixpkgs#272532]
### Describe the bug
I wanted to configure caddy with my own config file, setting `services.caddy.configFile = pkgs.writeTextDir "Caddyfile" ''…'';` as documented in the option.

### Steps To Reproduce
1. Use above snippet
2. Observe how caddy does not start, as `/etc/caddy/Caddyfile` does not exist.
3. look at the filesystem, observe there's a `/etc/caddy/caddy_config` directory with a `Caddyfile` inside, while `ExecStart` sets `--config /etc/caddy/Caddyfile`

Possibly introduced in c01874615be331a9f2832488ac56515bf1804197

### Notify maintainers

@Sohalt @Br1ght0ne @emilylange @techknowlogick 

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by emilylange** [comment]
Looks like a regression to me.
Will investigate further.

Thanks for your report and the commit you suggested that might have introduced it :)

**Comment by Sohalt** [comment]
Are you using an unmodified systemd unit?

The config file is expected to end up in `/etc/caddy/caddy_config/Caddyfile`
https://github.com/NixOS/nixpkgs/blob/ea5566eb3b278d405c7f2ecddae033ed2280ec8f/nixos/modules/services/web-servers/caddy/default.nix#L50-L52

and the service should refer to that path
https://github.com/NixOS/nixpkgs/blob/ea5566eb3b278d405c7f2ecddae033ed2280ec8f/nixos/modules/services/web-servers/caddy/default.nix#L360-L364

**Comment by Sohalt** [comment]
I don't think it's that commit. Can you try `pkgs.writeTextFile`?

**Comment by emilylange** [comment]
I am pretty sure this documentation error has been introduced all the way back in #153147.

The option docs simply switched from `pkgs.writeText` to `pkgs.writeTextDir`, but forgot to wrap add a ` + /Caddyfile;` at the end.

And now we are somewhat stuck with behavior that expects it that way... sigh.

```diff
diff --git a/nixos/modules/services/web-servers/caddy/default.nix b/nixos/modules/services/web-servers/caddy/default.nix
index 497aa9ba956e..b4ccbb223a36 100644
--- a/nixos/modules/services/web-servers/caddy/default.nix
+++ b/nixos/modules/services/web-servers/caddy/default.nix
@@ -153,7 +153,7 @@ in
           root * /var/www/wordpress
           php_fastcgi unix//run/php/php-version-fpm.sock
           file_server
-        ''';
+        ''' + "/Caddyfile";
       '';
       description = lib.mdDoc ''
         Override the configuration file used by Caddy. By default,
```

**Comment by Sohalt** [comment]
I think this in combination with https://github.com/NixOS/nixpkgs/commit/1f0ac736b48dbeaed1b3da975d13246b4d3a886c introduced the breaking change, since we now always pass `/etc/caddy/caddy_config` as the `--config` arg, but expect a basename of `Caddyfile` to enable the `caddyfile` adapter.

**Comment by Sohalt** [comment]
It works if you do `"${pkgs.writeTextDir …}/Caddyfile"` as demonstrated here: https://github.com/NixOS/nixpkgs/blob/583eaca796b40d068136169b9e160214cc473ee5/nixos/tests/caddy.nix#L52-L56
But it's ugly.

**Comment by flokli** [comment]
I tried with `pkgs.writeTextFile` too, it doesn't work. I extended `nixosTests.caddy` with an example (at least I think I did), and it fails.

It's pushed to https://github.com/flokli/nixpkgs/tree/caddy-regression-test, if someone does a PR fixing the issue feel free to cherry-pick it in ;-)

**Comment by Sohalt** [comment]
With `writeTextFile` you also need to explicitly set `adapter = "caddyfile"`.

**Comment by Sohalt** [comment]
Fixed it in https://github.com/NixOS/nixpkgs/pull/272542/commits/f61af31abf3efbd1e6b7e34047b14ba1e54f2bf7

(when using `writeText`)

**Comment by flokli** [comment]
Can you also fix the writeTextDir case? That's what the documentation recommends to use...

**Comment by Sohalt** [comment]
I adjusted the documentation to use `writeText` (which it used to be before https://github.com/NixOS/nixpkgs/pull/153147)

**Comment by flokli** [comment]
I think fixing the writeTextDir case makes more sense. That's what people might have in their configs, and getting everyone to notice they now need to write it like this, and update configs is worse UX than fixing the usage that was documented.

Assuming it worked with 23.05 for example, but regressed with 23.11 again.

That is, if this ever worked like this before. If it was broken all the time, no one could have a working config so it doesn't matter much :laughing:

**Comment by emilylange** [comment]
Both `pkgs.writeText` and `pkgs.writeTextDir` (without `+ "/Caddyfile";`) have been broken since #153147, which landed in `22.11`.

**Comment by flokli** [comment]
Ok, in that case, it's been broken for an entire release, so this is not just a regression when upgrading to 23.11. then it's probably ok.
